MHA torch.jit.script fix for in_proj_weight = None

mikekgfb opened the pull request:
Summary: MHA fix to support in_proj_weight being None

Test Plan: sandcastle

Differential Revision: D43628206

davidberard98 reviewed (approved):
LGTM

vadimkantorov commented on torch/nn/modules/activation.py:
```diff
@@ -1133,28 +1152,29 @@ def forward(
             if torch.overrides.has_torch_function(tensor_args):
                 why_not_fast_path = "some Tensor argument has_torch_function"
-            elif not all([(x is None or x.is_cuda or 'cpu' in str(x.device)) for x in tensor_args]):
+            elif not all([_arg_cuda_or_cpu(x) for x in tensor_args]):
```
here square brackets are not strictly needed...

mikekgfb replied:
actually, for Torchscript they are needed because torchscript only supports a subset of Python features :(

vadimkantorov commented on torch/nn/modules/activation.py:
```diff
@@ -886,6 +886,23 @@ def extra_repr(self) -> str:
+    else:
+        return x.requires_grad
+
+    return True
```
this seems unreachable, right?

mikekgfb replied:
In principle, unless you ever run into a type checker that checks implicit returns regardless of reachability.  Putting the conditional into the else path is required to get type refinement for static torchscript typing.

Commit:
MHA torch.jit.script fix for in_proj_weight = None (#95653)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/95653

MHA fix to support in_proj_weight being None during scripting

Test Plan: sandcastle

Reviewed By: davidberard98

Differential Revision: D43628206

fbshipit-source-id: 471b4e19bab00e986ea6a875341cba3dadd9679a

cpuhrsch commented on torch/nn/modules/activation.py:
```diff
@@ -886,6 +886,24 @@ def extra_repr(self) -> str:
         return str(self.lambd)
 
 
+def _arg_cuda_or_cpu(x: Optional[torch.Tensor]) -> bool:
+    if x is None:
+        return True
+    else:
+        return x.is_cuda or 'cpu' in str(x.device)
+
+    return False
```
```suggestion
def _arg_cuda_or_cpu(x: Optional[torch.Tensor]) -> bool:
    if x is None:
        return True
    return x.is_cuda or 'cpu' in str(x.device)
```

mikekgfb replied:
Putting the conditional into the else path is required to get type refinement for static torchscript typing. 

Type refinement is only active for the duration of the if/else from what I understand, and then reverts to pre-if typing, i.e., it will complain that a NoneType (which is part of Optional[tensor]) does not have is_cuda as attribute.

cpuhrsch commented on torch/nn/modules/activation.py:
```diff
@@ -886,6 +886,24 @@ def extra_repr(self) -> str:
+    return False
+
+
+def _arg_requires_grad(x: Optional[torch.Tensor]) -> bool:
+    if x is None:
+        return False
+    else:
+        return x.requires_grad
+
+    return True
```
```suggestion
def _arg_requires_grad(x: Optional[torch.Tensor]) -> bool:
    if x is None:
        return False
    return x.requires_grad
```

mikekgfb replied:
Ditto, see above.

apivovarov replied:
What if we use the following code?
```
def _arg_requires_grad(x: Optional[torch.Tensor]) -> bool:
    result: bool = False
    if x is not None:
        result = x.requires_grad
    return result
```

cpuhrsch reviewed (approved):
See inline suggestions

mikekgfb commented:
@pytorchbot merge

mikekgfb commented:
@pytorchbot merge
